site/content/blog/2020-01-08-incorporating-golangci-lint.md

```diff
@@ -0,0 +1,140 @@
+---
+title: Incorporating GolangCI-Lint at Mattermost
+slug: incorporating-golangci-lint
+date: 2020-01-08
+categories:
+    - "go"
+    - “linting”
+author: Agniva De Sarker
+github: agnivade
+community: agnivade
+author_2: Ben Schumacher
+github_2: hanzei
+community_2: hanzei
+---
+
+
+At Mattermost, we have traditionally relied on the trusty `go vet` and `gofmt` checks for our CI runs. While that is great, there were a lot of other powerful linters out there which we were missing out on.
+
+Speaking of linters, the first name that inevitably comes up is [staticcheck](https://staticcheck.io/). It is a powerful metalinter with a whole slew of checks. But simply running staticcheck was not sufficient, because it missed out on other linters which did one little thing, but nevertheless were very powerful. A few popular ones are [ineffassign](https://github.com/gordonklaus/ineffassign), [errcheck](https://github.com/kisielk/errcheck), [unconvert](https://github.com/mdempsky/unconvert), [deadcode](https://github.com/remyoudompheng/go-misc/tree/master/deadcode), [structcheck](https://gitlab.com/opennota/check/tree/master/cmd/structcheck) amongst others.
+
+We needed a way to run all of these linters together and efficiently too. Gometalinter was a pioneer in this field by providing a way to run multiple linters together. Unfortunately, it did this by shelling out to the linters and running them, which made it very inefficient. To its defence, using all the linters as a library was also equally hard, because there wasn’t any standard API that all of them could use.
+
+But the introduction of the `golang.org/x/tools/go/analysis` package during the 1.12 cycle was a game-changer. It introduced a standard API for writing Go static analyzers, which allowed them to be easily shared with the rest of the ecosystem in a plug-and-play model. Finally the One Ring to rule them all!
+
+Unsurprisingly, not before too long, GolangCI-Lint came up on the horizon. It used the go/analysis API to load all the linters and ran them concurrently, leading to a drastic reduction in memory and improvement in speed; becoming the successor to Gometalinter which is deprecated now.
+
+The writing was on the wall. GolangCI-Lint was the right choice. Following is an account of how we integrated it into our codebase.
+
+
+#### Integrating GolangCI-Lint into our CI
+
+Right off the bat, there was a big question mark hanging over our heads. It would be very, _very_ painful to have to fix _all_ the issues in our codebase before enforcing it as a CI check, by which time, new issues would have creeped in. It would be a never-ending chase.
```

Comment:
```suggestion
Right off the bat, there was a big question mark hanging over our heads. It would be very, _very_ painful to have to fix _all_ the issues in our codebase before enforcing it as a CI check, by which time, new issues would have crept in. It would be a never-ending chase.
```